libsrc/vic20/tgi/vic20-hi.s

```diff
@@ -0,0 +1,1054 @@
+
+; Step to next row on screen.
+
+        lda     tmp2
+        clc
+        adc     #20
+        sta     tmp2
+        bcc     @L1
+        inc     tmp2+1
+@L1:    inx
+
+        cpx     #12
+        bne     @NEXT_ROW
+
+; Set up VIC.
+
+        ldx #5
+@L2:    clc
+        lda     $EDE4,x
+        adc     VICREGS,x
+        sta     $9000,x
+        dex
+        bpl     @L2
+
+        lda     $900F
+        and     #%00000111
+        ora     #9
+        sta     $900F
+
+; Reset user defined pattern.
+
+        lda     #$00
+        sta     PATTERN
+        sta     PATTERN+1
+
+; Done, reset the error code
+
+        lda     #TGI_ERR_OK
+        sta     ERROR
+        rts
+.endproc
+
+; ------------------------------------------------------------------------
+; DONE: Will be called to switch the graphics device back into text mode.
+; The graphics kernel will never call DONE when no graphics mode is active,
+; so there is no need to protect against that.
+;
+; Must set an error code: NO
+;
+
+.proc DONE
+        jmp    $E518        ; KERNAL VIC init.
+.endproc
+
+; ------------------------------------------------------------------------
+; GETERROR: Return the error code in A and clear it.
+
+.proc GETERROR
+        ldx     #TGI_ERR_OK
+        lda     ERROR
+        stx     ERROR
+        rts
+.endproc
+
+; ------------------------------------------------------------------------
+; CONTROL: Platform/driver specific entry point.
+;
+; Must set an error code: YES
+;
+
+.proc CONTROL
+
+; Set user defined pattern.
+
+        cmp     #TGI_IOCTL_VIC20_SET_PATTERN
+        bne     @INVALID_FUNC
+
+        lda     ptr1
+        sta     USERPATTERN
+        lda     ptr1+1
+        sta     USERPATTERN+1
+
+        lda     #TGI_ERR_OK
+        sta     ERROR
+        rts
+
+; Return with error code for invalid function index.
+
+@INVALID_FUNC:
+        lda     #TGI_ERR_INV_FUNC
+        sta     ERROR
+        rts
+.endproc
+
+; ------------------------------------------------------------------------
+; CLEAR: Clears the screen.
+;
+; Must set an error code: NO
+;
+
+.proc CLEAR
+        ldy     #$00
+@L1:    lda     #$00
+        sta     $1100,y
+        sta     $1200,y
+        sta     $1300,y
+        sta     $1400,y
+        sta     $1500,y
+        sta     $1600,y
+        sta     $1700,y
+        sta     $1800,y
+        sta     $1900,y
+        sta     $1A00,y
+        sta     $1B00,y
+        sta     $1C00,y
+        sta     $1D00,y
+        sta     $1E00,y
+        sta     $1F00,y
+        lda     #1
+        sta     $9400,y
+        sta     $9500,y
+        iny
+        bne     @L1
+        rts
+.endproc
+
+; ------------------------------------------------------------------------
+; SETVIEWPAGE: Set the visible page. Called with the new page in A (0..n).
+; The page number is already checked to be valid by the graphics kernel.
+;
+; Must set an error code: NO (will only be called if page ok)
+;
+
+.proc SETVIEWPAGE
+        rts
```

Comment:
How about fall through to SETDRAWPAGE?

Reply:
I have no idea. This is about all I'll contribute. Just a port of the C64 code.

Reply:
The `SETVIEWPAGE` and `SETDRAWPAGE` functions (in this driver) don't do anything.  Therefore, "polluks" is suggesting that you can save one byte ("every little byte counts") by changing line 432 into
```
;rts     ; fall through
```
Both functions will share the second `rts`.

Reply:
I'd suggest to wait with optimisations until the TGI drivers have been completed. At least in the CBM section. Fellow developers will probably welcome the stubs.